README.md
# NAME

WebService::SendBird - unofficial support for the Sendbird API

# SYNOPSIS

    use WebService::SendBird;

    my $api = WebService::SendBird->new(
        api_token => 'You_Api_Token_Here',
        app_id    => 'You_App_ID_Here',
    );

    my $user = $api->create_user(
        user_id     => 'my_chat_user_1',
        nickname    => 'pumpkin',
        profile_url => undef,
    );

    my $chat = $api->create_group_chat(
        user_ids => [ $user->user_id ],
    );

# DESCRIPTION

Basic implementation for SendBird Platform API client, which helps to create users and group chats.

More information at [Platform API Documentation](https://docs.sendbird.com/platform)

# METHODS

## new

Creates an instance of API client

- `api_token` - Master or Secondary API Token.
- `app_id` - Sendbird Application ID.
- `api_url` - URL to API end point. By default it will be generated from app\_id.
- `ua` - Custom http client for API requests, should have the same interface like [Mojo::UserAgent](https://metacpan.org/pod/Mojo%3A%3AUserAgent).
- `timeout` - request timeout, default value 15 seconds

## app\_id

Returns Application ID.

## api\_token

Returns API Token

## api\_url

Returns API endpoint url

## timeout

Return http request timeout value.

## ua

Return User Agent for http request.

## http\_headers

Returns headers for API request.

## request

Sends request to Sendbird API

## create\_user

Creates a user at SendBird

- `user_id` - Unique User Identifier
- `nickname` - User nickname
- `profile_url` - user profile url. Could be `undef` or empty.

Information about extra parameters could be found at [API Documentation](https://docs.sendbird.com/platform/user#3_create_a_user)

Method returns an instance of [WebService::SendBird::User](https://metacpan.org/pod/WebService%3A%3ASendBird%3A%3AUser)

## view\_user

Gets information about a user from SendBird

- `user_id` - Unique User Identifier

Information about extra parameters could be found at [API Documentation](https://docs.sendbird.com/platform/user#3_view_a_user)

Method returns an instance of [WebService::SendBird::User](https://metacpan.org/pod/WebService%3A%3ASendBird%3A%3AUser)

## create\_group\_chat

Creates a group chat room

Information about parameters could be found at [API Documentation](https://docs.sendbird.com/platform/group_channel#3_create_a_channel)

Method returns an instance of [WebService::SendBird::GroupChat](https://metacpan.org/pod/WebService%3A%3ASendBird%3A%3AGroupChat)

## view\_group\_chat

Gets information about a group chat from SendBird

- `channel_url` - Unique Chat Identifier

Information about parameters could be found at [API Documentation](https://docs.sendbird.com/platform/group_channel#3_view_a_channel)

Method returns an instance of [WebService::SendBird::GroupChat](https://metacpan.org/pod/WebService%3A%3ASendBird%3A%3AGroupChat)


Changes

Revision history for {{$dist->name}}

{{$NEXT}}

0.002     2022-12-20 10:26:30+00:00 UTC
    remove debug line

0.001     2022-11-13 08:15:13+00:00 UTC
    Initial release.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 7.64;

my %WriteMakefileArgs = (
  "ABSTRACT" => "client to work with SendBird API",
  "AUTHOR" => "DERIV <DERIV\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.64"
  },
  "DISTNAME" => "WebService-SendBird",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "WebService::SendBird",
  "PREREQ_PM" => {
    "Carp" => 0,
    "JSON::MaybeXS" => 0,
    "JSON::PP" => 0,
    "Mojo::URL" => 0,
    "Mojo::UserAgent" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Fatal" => 0,
    "Test::MockObject" => 0,
    "Test::MockObject::Extends" => 0,
    "Test::More" => "0.94"
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeXS" => 0,
  "JSON::PP" => 0,
  "Mojo::URL" => 0,
  "Mojo::UserAgent" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Fatal" => 0,
  "Test::MockObject" => 0,
  "Test::MockObject::Extends" => 0,
  "Test::More" => "0.94",
  "constant" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/WebService/SendBird/User.pm
package WebService::SendBird::User;

use strict;
use warnings;

use Carp;
use JSON::MaybeXS;

## VERSION

=head1 NAME

WebService::SendBird::User - SendBird User

=head1 SYNOPSIS

 use WebService::SendBird::User;

 my $user = WebService::SendBird::User->new(
     api_client => $api,
     user_id    => 'my_chat_user_1',
 );

 $user->update(nickname => 'cucumber');

 my $token_data = $user->issue_session_token;

=head1 DESCRIPTION

Class for SendBird User. Information about structure could be found at L<API Documentation|https://docs.sendbird.com/platform/user>

=cut

use constant REQUIRED_FIELDS => qw(
    api_client
    user_id
);

use constant OPTIONAL_FIELDS => qw(
    phone_number
    has_ever_logged_in
    session_tokens
    access_token
    discovery_keys
    is_online
    last_seen_at
    nickname
    profile_url
    metadata
);

{
    no strict 'refs';
    for my $field (REQUIRED_FIELDS, OPTIONAL_FIELDS) {
        *{ __PACKAGE__ . '::' . $field } = sub { shift->{$field} };
    }
}

=head2 new

Creates an instance of SendBird User

=over 4

=item * C<api_client> - SendBird API client L<WebService::SendBird>.

=item * C<user_id> - Unique User Identifier

=back

=cut

sub new {
    my ($cls, %params) = @_;

    my $self = +{};
    $self->{$_} = delete $params{$_} or Carp::croak "$_ is missed" for (REQUIRED_FIELDS);

    $self->{$_} = delete $params{$_} for (OPTIONAL_FIELDS);

    return bless $self, $cls;
}

=head2 Getters

=over 4

=item * C<api_client>

=item * C<user_id>

=item * C<phone_number>

=item * C<has_ever_logged_in>

=item * C<session_tokens>

=item * C<access_token>

=item * C<discovery_keys>

=item * C<is_online>

=item * C<last_seen_at>

=item * C<nickname>

=item * C<profile_url>

=item * C<metadata>

=back

=cut

=head2 update

Updates the user at SendBird API

Information about parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/user#3_update_a_user>

=cut

sub update {
    my ($self, %params) = @_;

    my $res = $self->api_client->request(PUT => 'users/' . $self->user_id, \%params);

    $self->{$_} = $res->{$_} for (OPTIONAL_FIELDS);

    return $self;
}

=head2 issue_session_token

Issues new session token and returns hash ref with token and expiration time of this token.

=cut

sub issue_session_token {
    my ($self) = @_;

    $self->update(issue_session_token => JSON::MaybeXS::true);

    my $tokens = $self->session_tokens // [];

    my ($latest_token) = sort { $b->{expires_at} <=> $a->{expires_at} } @$tokens;

    return $latest_token;
}

1;


lib/WebService/SendBird/Message.pm
package WebService::SendBird::Message;

use strict;
use warnings;

use Carp;
use JSON::PP;

use WebService::SendBird::User;
=head1 NAME

WebService::SendBird::Message - Base class for SendBird Message

=head1 SYNOPSIS

 use WebService::SendBird::User;

 my $user = WebService::SendBird::User->new(
     api_client => $api,
     user_id    => 'my_chat_user_1',
 );

 $user->update(nickname => 'cucumber');

 my $token_data = $user->issue_session_token;

=head1 DESCRIPTION

Class for SendBird User. Information about structure could be found at L<API Documentation|https://docs.sendbird.com/platform/user>

=cut

use constant REQUIRED_FIELDS => qw(
    api_client
    type
);

use constant OPTIONAL_FIELDS => qw(
    custom_type
    data
    send_push
    mention_type
    mentioned_user_ids
    is_silent
    sorted_metaarray
    created_at
    dedup_id
    user_id
    message
    mark_as_read
    file
    url
    file_name
    file_size
    file_type
    thumbnails
    require_auth
    user
    mentioned_users
);

{
    no strict 'refs';
    for my $field (REQUIRED_FIELDS, OPTIONAL_FIELDS) {
        *{ __PACKAGE__ . '::' . $field } = sub { shift->{$field} };
    }
}

=head2 new

Creates an instance of SendBird User

=over 4

=item * C<api_client> - SendBird API client L<WebService::SendBird>.

=item * C<user_id> - Unique User Identifier

=back

=cut

sub new {
    my ($cls, %params) = @_;

    my $self = +{};
    $self->{$_} = delete $params{$_} or Carp::croak "$_ is missed" for (REQUIRED_FIELDS);

    $self->{$_} = delete $params{$_} for (OPTIONAL_FIELDS);

    if ( $self->{user} ) {
        $self->{user} = WebService::SendBird::User->new(%{ $self->{user} }, api_client => $self->{api_client});
    }

    $self->{mentioned_users} //= [];
    my @obj_mentions = map { WebService::SendBird::User->new(%$_, api_client => $self->{api_client}) }  @{$self->{mentioned_users}};
    $self->{mentioned_users} = \@obj_mentions;

    return bless $self, $cls;
}

=head2 Getters

=over 4

=item * C<api_client>

=item * C<user_id>

=item * C<phone_number>

=item * C<has_ever_logged_in>

=item * C<session_tokens>

=item * C<access_token>

=item * C<discovery_keys>

=item * C<is_online>

=item * C<last_seen_at>

=item * C<nickname>

=item * C<profile_url>

=item * C<metadata>

=back

=cut

1;


lib/WebService/SendBird/GroupChat.pm
package WebService::SendBird::GroupChat;

use strict;
use warnings;

use Carp;
use JSON::MaybeXS ();
use WebService::SendBird::User;

## VERSION

=head1 NAME

WebService::SendBird::GroupChat - SendBird Group Chat

=head1 SYNOPSIS

 use WebService::SendBird::Group;

 my $chat = WebService::SendBird::User->new(
     api_client  => $api,
     channel_url => 'chat_room_1',
 );

 $chat->update(name => 'Gossips');

=head1 DESCRIPTION

Class for SendBird Group Channel. Information about structure could be found at L<API Documentation|https://docs.sendbird.com/platform/group_channel>

=cut

use constant REQUIRED_FIELDS => qw(
    api_client
    channel_url
);

use constant OPTIONAL_FIELDS => qw(
    is_broadcast
    name
    is_access_code_required
    is_super
    joined_member_count
    is_public
    cover_url
    unread_mention_count
    is_created
    is_distinct
    is_ephemeral
    freeze
    data
    is_discoverable
    last_message
    sms_fallback_interval_sec
    custom_type
    unread_message_count
    created_at
    member_count
    sms_fallback_enabled
    max_length_message
    members
);

{
    no strict 'refs';
    for my $field (REQUIRED_FIELDS, OPTIONAL_FIELDS) {
        *{ __PACKAGE__ . '::' . $field } = sub { shift->{$field} };
    }
}

=head2 new

Creates an instance of SendBird Group Chat

=over 4

=item * C<api_client> - SendBird API client L<WebService::SendBird>.

=item * C<channel_url> - Unique Channel Identifier

=back

=cut

sub new {
    my ($cls, %params) = @_;

    my $self = +{};
    $self->{$_} = delete $params{$_} or Carp::croak "$_ is missed" for (REQUIRED_FIELDS);

    $self->{$_} = delete $params{$_} for (OPTIONAL_FIELDS);

    $self->{members} //= [];
    my @obj_members = map { WebService::SendBird::User->new(%$_, api_client => $self->{api_client}) }  @{$self->{members}};
    $self->{members} = \@obj_members;

    return bless $self, $cls;
}

=head2 Getters

=over 4

=item * C<api_client>

=item * C<channel_url>

=item * C<is_broadcast>

=item * C<name>

=item * C<is_access_code_required>

=item * C<is_super>

=item * C<joined_member_count>

=item * C<is_public>

=item * C<cover_url>

=item * C<unread_mention_count>

=item * C<is_created>

=item * C<is_distinct>

=item * C<is_ephemeral>

=item * C<freeze>

=item * C<data>

=item * C<is_discoverable>

=item * C<last_message>

=item * C<sms_fallback_interval_sec>

=item * C<custom_type>

=item * C<unread_message_count>

=item * C<created_at>

=item * C<member_count>

=item * C<sms_fallback_enabled>

=item * C<max_length_message>

=item * C<members>

=back

=cut

=head2 update

Updates the group channel at SendBird API

Information about parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/group_channel#3_update_a_channel>

=cut

sub update {
    my ($self, %params) = @_;

    my $res = $self->api_client->request(PUT => 'group_channels/' . $self->channel_url, \%params);

    $self->{$_} = $res->{$_} for qw(OPTIONAL_FIELDS);

    return $self
}

=head2 set_freeze

Freeze or unfreeze the channel.

See L<https://sendbird.com/docs/chat/v3/platform-api/guides/group-channel#2-freeze-a-channel>.

=cut

sub set_freeze {
    my ($self, $freeze) = @_;
    
    $freeze = $freeze ? JSON::MaybeXS::true : JSON::MaybeXS::false;

    my $res = $self->api_client->request(PUT => 'group_channels/' . $self->channel_url. '/freeze', { freeze => $freeze } );

    $self->{freeze} = $res->{freeze};
    return $self
}

1;


lib/WebService/SendBird.pm
package WebService::SendBird;

use strict;
use warnings;

use Mojo::UserAgent;
use Mojo::URL;

use WebService::SendBird::User;
use WebService::SendBird::GroupChat;

use Carp qw();

our $VERSION = '0.003';

# ABSTRACT: client to work with SendBird API

=head1 NAME

WebService::SendBird - unofficial support for the Sendbird API

=head1 SYNOPSIS

 use WebService::SendBird;

 my $api = WebService::SendBird->new(
     api_token => 'You_Api_Token_Here',
     app_id    => 'You_App_ID_Here',
 );

 my $user = $api->create_user(
     user_id     => 'my_chat_user_1',
     nickname    => 'pumpkin',
     profile_url => undef,
 );

 my $chat = $api->create_group_chat(
     user_ids => [ $user->user_id ],
 );

=head1 DESCRIPTION

Basic implementation for SendBird Platform API client, which helps to create users and group chats.

More information at L<Platform API Documentation|https://docs.sendbird.com/platform>

=cut

use constant DEFAULT_API_URL_TEMPLATE => 'https://api-%s.sendbird.com/v3';
use constant DEFAULT_REQUEST_TIMEOUT  => 15;

=head1 METHODS

=head2 new

Creates an instance of API client

=over 4

=item * C<api_token> - Master or Secondary API Token.

=item * C<app_id> - Sendbird Application ID.

=item * C<api_url> - URL to API end point. By default it will be generated from app_id.

=item * C<ua> - Custom http client for API requests, should have the same interface like L<Mojo::UserAgent>.

=item * C<timeout> - request timeout, default value 15 seconds

=back

=cut


sub new {
    my ($cls, %params) = @_;

    Carp::croak('Missing required argument: api_token') unless $params{api_token};
    Carp::croak('Missing required argument: app_id or api_url') unless $params{app_id} || $params{api_url};

    my $self = +{
        api_token => $params{api_token},
        $params{ua}      ? (ua      => $params{ua})                      : (),
        $params{app_id}  ? (app_id  => $params{app_id})                  : (),
        $params{api_url} ? (api_url => Mojo::URL->new($params{api_url})) : (),
    };

    return bless $self, $cls;
}


=head2 app_id

Returns Application ID.

=cut

sub app_id {return shift->{app_id} }

=head2 api_token

Returns API Token

=cut

sub api_token {return shift->{api_token} }

=head2 api_url

Returns API endpoint url

=cut

sub api_url {
    my $self = shift;

    $self->{api_url} //= Mojo::URL->new(sprintf(DEFAULT_API_URL_TEMPLATE, $self->app_id));

    return $self->{api_url};
}

=head2 timeout

Return http request timeout value.

=cut

sub timeout {return shift->{timeout} || DEFAULT_REQUEST_TIMEOUT }

=head2 ua

Return User Agent for http request.

=cut

sub ua {
    my $self = shift;

    return $self->{ua} if $self->{ua};

    my $ua = Mojo::UserAgent->new();

    $ua->inactivity_timeout($self->timeout);
    $ua->proxy->detect;

    $self->{ua} = $ua;
    return $self->{ua};
}

=head2 http_headers

Returns headers for API request.

=cut

sub http_headers {
    my $self = shift;

    return {
        'Content-Type' => 'application/json, charset=utf8',
        'Api-Token'    => $self->api_token,
    }
}


=head2 request

Sends request to Sendbird API

=cut

sub request {
    my ($self, $method, $path, $params) = @_;

    my $resp = $self->ua->start(
        $self->ua->build_tx(
            $method,
            $self->_url_for($path),
            $self->http_headers,
            uc($method) eq 'GET' ? (form => $params) : (json => $params),
        )
    );

    if ( $resp->result->is_error ) {
        my $details = $resp->result->{content}{asset}{content} // $resp->result->message;
        Carp::croak('Fail to make request to SB API: '.$details );
    }

    my $data;
    eval {
        $data = $resp->result->json;
        1;
    } or do {
        Carp::croak('Fail to parse response from SB API: ' . $@);
    };

    Carp::croak('Fail to make request to SB API: ' . $data->{message}) if $data->{error};

    return $data;
}

=head2 create_user

Creates a user at SendBird

=over 4

=item * C<user_id> - Unique User Identifier

=item * C<nickname> - User nickname

=item * C<profile_url> - user profile url. Could be C<undef> or empty.

=back

Information about extra parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/user#3_create_a_user>

Method returns an instance of L<WebService::SendBird::User>

=cut

sub create_user {
    my ($self, %params) = @_;

    Carp::croak('profile_url is missed') unless exists $params{profile_url};
    $params{$_} or Carp::croak("$_ is missed") for (qw(user_id nickname));

    my $resp = $self->request(POST => 'users', \%params);

    return WebService::SendBird::User->new(%$resp, api_client => $self);
}


=head2 view_user

Gets information about a user from SendBird

=over 4

=item * C<user_id> - Unique User Identifier

=back

Information about extra parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/user#3_view_a_user>

Method returns an instance of L<WebService::SendBird::User>

=cut

sub view_user {
    my ($self, %params) = @_;

    my $user_id = delete $params{user_id} or Carp::croak('user_id is missed');

    my $resp = $self->request(GET => "users/$user_id", \%params);

    return WebService::SendBird::User->new(%$resp, api_client => $self);
}

=head2 create_group_chat

Creates a group chat room

Information about parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/group_channel#3_create_a_channel>

Method returns an instance of L<WebService::SendBird::GroupChat>

=cut

sub create_group_chat {
    my ($self, %params) = @_;

    my $resp = $self->request(POST => "group_channels", \%params);

    return WebService::SendBird::GroupChat->new(%$resp, api_client => $self);
}


=head2 view_group_chat

Gets information about a group chat from SendBird

=over 4

=item * C<channel_url> - Unique Chat Identifier

=back

Information about parameters could be found at L<API Documentation|https://docs.sendbird.com/platform/group_channel#3_view_a_channel>

Method returns an instance of L<WebService::SendBird::GroupChat>

=cut

sub view_group_chat {
    my ($self, %params) = @_;
    my $channel_url = delete $params{channel_url} or Carp::croak('channel_url is missed');

    my $resp = $self->request(GET => "group_channels/$channel_url", \%params);

    return WebService::SendBird::GroupChat->new(%$resp, api_client => $self);
}

#Private methods

#Returns full URL for requested path
sub _url_for {
    my ($self, $path) = @_;

    return join q{/} => ($self->api_url, $path);
}


1;
